databuilder/query_model.py

```diff
@@ -1,452 +1,356 @@
+        ...
 
-    def __invert__(self):
-        return boolean_comparator(self, negated=True)
+    class NE(BinaryFunction):
+        ...
 
-    def __hash__(self):
-        return id(self)
+    class LT(BinaryFunction):
+        ...
 
+    class LE(BinaryFunction):
+        ...
 
-@dataclass(frozen=True, eq=False, order=False)
-class ValueFromRow(Value):
-    source: Any
-    column: Any
+    class GT(BinaryFunction):
+        ...
 
-    def _get_referenced_nodes(self):
-        return (self.source,)
+    class GE(BinaryFunction):
+        ...
 
+    # Boolean
+    class And(BinaryFunction):
+        ...
 
-@dataclass(frozen=True, eq=False, order=False)
-class ValueFromAggregate(Value):
-    source: RowFromAggregate
-    column: Any
+    class Or(BinaryFunction):
+        ...
 
-    def _get_referenced_nodes(self):
-        return (self.source,)
+    class Not(UnaryFunction):
+        ...
 
+    # Null handling
+    class IsNull(UnaryFunction):
+        ...
 
-def categorise(mapping, default=None):
-    mapping = {
-        key: boolean_comparator(value) if isinstance(value, Value) else value
-        for key, value in mapping.items()
-    }
-    return ValueFromCategory(mapping, default)
+    # Arithmetic
+    class Add(BinaryFunction):
+        ...
 
+    class Subtract(BinaryFunction):
+        ...
 
-@dataclass(frozen=True, eq=False, order=False)
-class ValueFromCategory(Value):
-    definitions: dict
-    default: str | int | float | None
+    # Dates
+    class RoundToFirstOfMonth(UnaryFunction):
+        ...
 
-    def _get_referenced_nodes(self):
-        nodes = ()
-        for value in self.definitions.values():
-            assert isinstance(value, QueryNode)
-            nodes += (value,)
-        return nodes
+    class RoundToFirstOfYear(UnaryFunction):
+        ...
 
+    class DateAdd(BinaryFunction):
+        ...
 
-@dataclass(frozen=True)
-class Codelist(QueryNode):
-    codes: tuple
-    system: str
-    has_categories: bool = False
+    class DateSubtract(BinaryFunction):
+        ...
 
-    def _get_referenced_nodes(self):
-        return ()
+    class DateDifference(Column):
+        start: Union[Column, datetime.date]
+        end: Union[Column, datetime.date]
+        units: str
 
-    def __post_init__(self):
-        if self.has_categories:
-            raise NotImplementedError("Categorised codelists are currently unsupported")
+    # Containment is a special case: its right-hand side must be something vector-like i.e.
+    # something containing multiple values. To build a column whose values are vectors,
+    # use the `CombineAsSet` aggregation.
+    class In(Column):
+        lhs: Column
+        rhs: Union[Column, tuple[Value], Codelist]
 
-    def __repr__(self):
-        if len(self.codes) > 5:
-            codes = self.codes[:5] + ("...",)
-        else:
-            codes = self.codes
-        return f"Codelist(system={self.system}, codes={codes})"
+    class NotIn(In):
+        ...
```

Comment:
Do we need `NotIn` rather than `Not(In())`?

Reply:
We do not. Done.